service-base/src/main/java/org/eclipse/hono/service/http/TenantTraceSamplingHandler.java

```diff
@@ -0,0 +1,188 @@
+                .setHandler(ar -> ctx.next());
+    }
+
+    @Override
+    public void handle(final RoutingContext ctx) {
+        getTenantViaCert(ctx)
+                .recover(thr -> getTenantFromAuthHeader(ctx))
+                .compose(tenantObjectWithAuthId -> applyTraceSamplingPriority(ctx, tenantObjectWithAuthId))
+                .setHandler(ar -> ctx.next());
+    }
+
+    private Future<TenantObjectWithAuthId> getTenantViaCert(final RoutingContext ctx) {
+        if (ctx.request().isSSL()) {
+            final X509Certificate deviceCert = getX509Cert(ctx.request().sslSession());
+            if (deviceCert != null) {
+                final X500Principal x500Principal = deviceCert.getIssuerX500Principal();
+                return tenantClientFactory.getOrCreateTenantClient()
+                        .compose(tenantClient -> tenantClient.get(x500Principal))
+                        .map(tenantObject -> new TenantObjectWithAuthId(tenantObject, null));
+            }
+        }
+        return Future.failedFuture("no cert found");
+    }
+
+    private X509Certificate getX509Cert(final SSLSession sslSession) {
+        try {
+            final Certificate[] path = sslSession.getPeerCertificates();
+            if (path.length > 0 && path[0] instanceof X509Certificate) {
+                return (X509Certificate) path[0];
+            }
+        } catch (final SSLPeerUnverifiedException e) {
+            LOG.debug("certificate chain cannot be read: {}", e.getMessage());
+        }
+        return null;
+    }
+
+    private Future<TenantObjectWithAuthId> getTenantFromAuthHeader(final RoutingContext ctx) {
+        final String authorizationHeader = ctx.request().headers().get(HttpHeaders.AUTHORIZATION);
+        if (authorizationHeader == null) {
+            return Future.failedFuture("no auth header");
+        }
+        String tenantIdTmp = null;
+        String authIdTmp = null;
+        try {
+            final int idx = authorizationHeader.indexOf(' ');
+            if (idx > 0 && "Basic".equalsIgnoreCase(authorizationHeader.substring(0, idx))) {
+                final String authorization = authorizationHeader.substring(idx + 1);
+                final String decoded = new String(Base64.getDecoder().decode(authorization));
+                final int colonIdx = decoded.indexOf(":");
+                final String userName = colonIdx != -1 ? decoded.substring(0, colonIdx) : decoded;
+                if (config.isSingleTenant()) {
+                    tenantIdTmp = Constants.DEFAULT_TENANT;
+                    authIdTmp = userName;
+                } else {
+                    // userName is <authId>@<tenantId>
+                    final String[] userComponents = userName.split("@", 2);
+                    if (userComponents.length == 2) {
+                        tenantIdTmp = userComponents[1];
+                        authIdTmp = userComponents[0];
+                    }
+                }
```

Comment:
Note: The logic above:
````
                if (config.isSingleTenant()) {
                    tenantIdTmp = Constants.DEFAULT_TENANT;
                    authIdTmp = userName;
                } else {
                    // userName is <authId>@<tenantId>
                    final String[] userComponents = userName.split("@", 2);
                    if (userComponents.length == 2) {
                        tenantIdTmp = userComponents[1];
                        authIdTmp = userComponents[0];
                    }
                }
````
comes from `UsernamePasswordAuthProvider#getCredentials` (and the `UsernamePasswordCredentials.create` call inside).
I've considered using the `usernamePasswordAuthProvider` instance of the `VertxBasedHttpProtocolAdapter` here, to call `getCredentials` in order to determine tenant/authId from the userName.
But, `getCredentials` belongs to the `CredentialsApiAuthProvider` class. The `VertxBasedHttpProtocolAdapter.usernamePasswordAuthProvider` instance however is of the `HonoClientBasedAuthProvider` type, which is the parent class of `CredentialsApiAuthProvider`. So, using `getCredentials` doesn't work here in all cases.

How about we require the `VertxBasedHttpProtocolAdapter` auth providers to be `CredentialsApiAuthProvider` instances (instead of the more generic type)? Aren't adapters supposed to use the Credentials API to fetch credentials anyway?

Reply:
In any case, I can put the above logic in a separate class so that a separate implementation may be used in the case that the `VertxBasedHttpProtocolAdapter` instance uses a completely different custom auth provider to extract tenant/auth-id from the `RoutingContext`.

I would do that in a subseqent commit then, along with some refactoring enabling code sharing with the upcoming trace-sampling-mode implementations for the other adapters.

Reply:
I've amended the commit here with the above mentioned refactoring.